Fix Client Auth IT

robobario opened the pull request:
Also fixes client authz IT

tombentley committed:
kroxylicious-authorizer-acl: Implement the Authorizer API using ACLs

Signed-off-by: Tom Bentley <tbentley@redhat.com>

robobario committed:
Enforce AddPartitionsToTxn authorization

Users need write permissions for the topics they are trying to add

This RPC is a little different in that the broker will fail ALL of the
topic/partitions in the request if ANY of them fails for whatever
reason. To retain this atomicity we shortcircuit respond immediately if
any topic is denied.

Note that we only implement versions 0-3 because versions after that are
intended to be called by the broker only, not clients. The requests
fail if the user does not have Cluster permissions. I think we can focus
first on the RPCs that may be received from older clients. See also
KIP-890 which deprecated clients sending AddPartitionsToTxn.

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

tombentley committed:
kroxylicious-filters: Use SubjectBuilder in SASL inspector (and IT's terminators)

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
kroxylicious-authorization: Add a filter which applies an Authorizer

Signed-off-by: Tom Bentley <tbentley@redhat.com>

robobario committed:
Metadata enforcement for topic id requests

* If a request exclusively uses topicIds in the topics list, then we can
treat it as an idempotent request. If any of the topicIds are missing,
then the broker will respond with UNKNOWN_TOPIC_ID for that topic.
* In the unknown topic id case we need to handle a null topic name.

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

robobario committed:
Fixup merge issue

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

robobario committed:
Authorize DescribeTopicPartitions

- if the client is requesting all topics, we must remove any unauthorized
topics from the response and update the topicAuthorizedOps bitmap on
topics the user is allowed to describe.
- if the client requests specific topics, then we remove unauthorized
topics before forwarding. We augment removed topic responses back into
the response and update topic authorized ops on topics that were
forwarded.

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

tombentley committed:
kroxylicious-api+runtime: Add SASL subject builder

* Subjects can arise as a result of SASL authentication.
* We add a plugin interface, SaslSubjectBuilderService, following the established convention.
* Because the Principal subtypes are open, the builder provide a plugin point for adding principals to the Subject which are unknown to the kroxylicious-api or -runtime modules.
* We also want to allow the possilibity of using information looked up from external systems such as LDAP/ActiveDirectory or an OAuth token introspection endpoint. Thus the builder has an asynchronous return type.
* The runtime adds a DefaultSubjectBuilder implementation which can be used for SASL-based Subject building.

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
kroxylicious-api+runtime: TLS subject builders

* Subjects can arise as a result of TLS client authentication.
* We add a plugin interface, TransportSubjectBuilderService, following the established convention.
* Because the Principal subtypes are open, the builders provide plugin points for adding principals to the Subject which are unknown to the kroxylicious-api or -runtime modules.
* We also want to allow the possilibity of using information looked up from external systems such as LDAP/ActiveDirectory or an OAuth token introspection endpoint. Thus the builders have an asynchronous return type.
* Allows the DefaultSubjectBuilder to also work for TLS-based subjects

Signed-off-by: Tom Bentley <tbentley@redhat.com>

robobario committed:
Fix client authz IT

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

robobario commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/filter/authorization/ClientAuthzIT.java:
```diff
@@ -558,7 +564,8 @@ public Outcome<ConsumerGroupMetadata> groupMetadata() {
                 outcome = Fail.of(e);
             }
-            return trace("groupMetadata", outcome.map(m -> new ConsumerGroupMetadata(m.groupId(), m.generationId(), "CLOBBERED", m.groupInstanceId())));
+            trace("groupMetadata", outcome.map(m -> new ConsumerGroupMetadata(m.groupId(), m.generationId(), "CLOBBERED", m.groupInstanceId())));
```
this was breaking things, we returned ConsumerGroupMetadata with clobbered bits, and then fed it back into the kafka APIs

tombentley commented on kroxylicious-filters/kroxylicious-authorization/src/main/java/io/kroxylicious/filter/authorization/AuthorizationFilter.java:
```diff
@@ -82,6 +82,9 @@ CompletionStage<ResponseFilterResult> onResponse(ResponseHeaderData header, ApiV
         apiEnforcement.put(ApiKeys.SYNC_GROUP, new Passthrough<>(0, 5));
+        apiEnforcement.put(ApiKeys.DESCRIBE_CONFIGS, new Passthrough<>(1, 4));
+        apiEnforcement.put(ApiKeys.ALTER_CONFIGS, new Passthrough<>(0, 2));
+        apiEnforcement.put(ApiKeys.INCREMENTAL_ALTER_CONFIGS, new Passthrough<>(0, 1));
```
A `ConfigResource` can name a topic, and should be subject to authz (see https://docs.confluent.io/platform/current/security/authorization/acls/overview.html#topic-resource-type-operations for example). So I don't think it's as simple as adding Passthrough.

robobario replied:
done

robobario commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/filter/authorization/ClientAuthzIT.java:
```diff
@@ -688,13 +695,18 @@ public void run(ClientFactory clientFactory) throws InterruptedException {
                             ConsumerConfig.CLIENT_ID_CONFIG, "consumer2",
-                            ConsumerConfig.GROUP_ID_CONFIG, GROUP_1,
-                            ConsumerConfig.GROUP_INSTANCE_ID_CONFIG, GROUP_1 + "-instance-1",
+                            ConsumerConfig.GROUP_ID_CONFIG, GROUP_2,
```
I figured this second consumer was easier to reason about by giving it a new group, so you know that the earlier consumer hasn't impacted it

robobario committed:
Enforce describe configs for topics

Require DESCRIBE_CONFIGS permission on a topic to describe its configs.

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

tombentley committed:
Revert "Amend javadoc"

This reverts commit f758cd0ce588eeb5ad279b00494a2f17c24513e6.

tombentley committed:
Support TransportSubjectBuilder asynchrony in the runtime

Signed-off-by: Tom Bentley <tbentley@redhat.com>

robobario committed:
Enforce alter configs authorization

User must be permitted to ALTER_CONFIGS for topic. If they aren't then
we do not forward on the config alteration and augment in an error code
on the response.

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

robobario committed:
Enforce incremental alter configs authorization

Users must be permitted to ALTER_CONFIGS for a topic. If we encounter
requests to alter configs for unauthorized topics we do not forward
those requests upstream and we augment in an error response.

Signed-off-by: Robert Young <robertyoungnz@gmail.com>

tombentley committed:
format

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
MyTransportSubjectBuilderService

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
Formatting

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
Imports

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
ACL Authorizer use exception

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
Fix ACL authz for renamed api

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
Fix authz filter for renamed api

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
Remove callers of method

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
Remove Unit

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
Revert "Remove Unit"

This reverts commit a12a5480a8c4b73f7d8876f386c8cf03833144cb.

tombentley committed:
Javadoc

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
comments

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
Move PrincipalFactor and UserFactory to the api module

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley committed:
messy

Signed-off-by: Tom Bentley <tbentley@redhat.com>

tombentley commented:
Merged directly.
